### PR TITLE
AF-212 new location style support

### PR DESCRIPTION
**Pull request by sandlerr**
:bread:

/cc @zendesk/vegemite 

### References
 - Jira link: https://zendesk.atlassian.net/browse/AF-212
 - related pr: https://github.com/zendesk/zendesk_app_framework/pull/521

### CODE REVIEW

File: spec/package_spec.rb [hunk 1]
```diff
@@ -1,5 +1,5 @@
 require 'spec_helper'
-require 'fileutils'
```
Comment (sandlerr) [inline spec/package_spec.rb]: revert this change?

File: lib/zendesk_apps_support/package.rb [hunk 2]
```diff
@@ -65,7 +65,7 @@ def assets
     end
 
     def path_to(file)
-      File.join(@root, file)
+      File.join(root, file)
```
Comment (ocke) [inline lib/zendesk_apps_support/package.rb]: Should we change @root -> root everywhere then?

Comment (sandlerr) [inline lib/zendesk_apps_support/package.rb]: done boss

File: .ruby-version [hunk 3]
```diff
@@ -1 +1 @@
-2.2.3
+2.3.0
```
Comment (ocke) [inline .ruby-version]: 2.3.0 doesn't exist for me, 2.3.0-dev does

Comment (adammw) [inline .ruby-version]: you'll need to update your `ruby-build`

Comment (sandlerr) [inline .ruby-version]: Olaf,  `cd ~/.rbenv/plugins/ruby-build && git pull && cd - && rbenv install 2.3.0`

**Comment by sandlerr** [comment]
@ocke @danielbreves :+1:?

File: lib/zendesk_apps_support/assets/src.js.erb [hunk 4]
```diff
@@ -1,32 +1,35 @@
...
-      },
-      translations: <%= translations.to_json %>,
-      templates: <%= templates.to_json %>,
-      frameworkVersion: <%= framework_version.to_json %>
-    });
-
-  ZendeskApps[<%= name.to_json %>] = app;
 }
+<% end %>
+var app = ZendeskApps.defineApp(<%= iframe_only ? 'null' : 'source' %>)
```
Comment (danielbreves) [inline lib/zendesk_apps_support/assets/src.js.erb]: Why `null` instead of no arguments?

Comment (danielbreves) [inline lib/zendesk_apps_support/assets/src.js.erb]: Perhaps we could define `source` at the top and leave it `undefined` for `iframe_only`?

File: lib/zendesk_apps_support/assets/src.js.erb [hunk 5]
```diff
@@ -1,32 +1,35 @@
+<% unless iframe_only %>
```
Comment (danielbreves) [inline lib/zendesk_apps_support/assets/src.js.erb]: Would prefer `if legacy`, but no big deal.

File: spec/fixtures/app_en.js [hunk 6]
```diff
@@ -1,23 +1,23 @@
 with( ZendeskApps.AppScope.create() ) {
```
Comment (danielbreves) [inline spec/fixtures/app_en.js]: Should we rename this file to `legacy_app_en.js`?

File: spec/fixtures/app_nl.js [hunk 7]
```diff
@@ -1,23 +1,23 @@
 with( ZendeskApps.AppScope.create() ) {
```
Comment (danielbreves) [inline spec/fixtures/app_nl.js]: Same as above.

File: lib/zendesk_apps_support/assets/src.js.erb (hunk 5 above)
Comment (adammw) [inline lib/zendesk_apps_support/assets/src.js.erb]: I think roman and I thought there'd be too much "baggage" in the word "legacy", especially if we change again or add new features, so a feature test made more sense.

Comment (danielbreves) [inline lib/zendesk_apps_support/assets/src.js.erb]: Makes sense :+1:

File: lib/zendesk_apps_support/assets/src.js.erb (hunk 4 above)
Comment (adammw) [inline lib/zendesk_apps_support/assets/src.js.erb]: It felt like a bit too much variable pollution/shadowing, as I wasn't sure if this was wrapped within an IIFE or not. null seemed like a more explicit argument/sentinel value as opposed to just leaving it undefined.

**Comment by sandlerr** [comment]
@ocke @danielbreves :+1:?

**Comment by danielbreves** [comment]
:+1:

**Comment by sandlerr** [comment]
@adammw I'm happy to merge this, are you :+1: also?

File: lib/zendesk_apps_support/validations/manifest.rb [hunk 8]
```diff
@@ -51,6 +51,14 @@ def ban_framework_version(manifest)
           ValidationError.new(:no_framework_version_required) unless manifest['frameworkVersion'].nil?
         end
 
+        # TODO: decide if we need translations in iframe apps
```
Comment (adammw) [inline lib/zendesk_apps_support/validations/manifest.rb]: can we do the TODO before merging?

File: lib/zendesk_apps_support/validations/source.rb [hunk 9]
```diff
@@ -26,17 +26,23 @@ def call(package)
           files = package.js_files
           app   = files.find { |file| file.relative_path == 'app.js' }
 
-          if package.manifest_json['requirementsOnly']
+          if package_needs_app_js?(package)
+            return [ ValidationError.new(:missing_source) ] unless app
+          else
             return app ? [ValidationError.new(:no_app_js_required)] : []
```
Comment (adammw) [inline lib/zendesk_apps_support/validations/source.rb]: Wouldn't this cause two errors for IFO apps for having app.js?

Comment (adammw) [inline lib/zendesk_apps_support/validations/source.rb]: I mean I assume this also covers requirementsOnly though

File: lib/zendesk_apps_support/validations/source.rb [hunk 10]
```diff
@@ -26,17 +26,23 @@ def call(package)
...
-          return [ValidationError.new(:missing_source)] unless app
-
           jshint_errors(files).flatten!
         end
 
         private
 
+        def package_needs_app_js?(package)
+          return false if package.manifest_json['requirementsOnly']
+          return true if package.locations.values.flat_map(&:values).any? { |l| l == '_legacy' }
```
Comment (adammw) [inline lib/zendesk_apps_support/validations/source.rb]: Use `iframe_only?` here? Also how come that condition looks so much more compact than the function body of `iframe_only?` - can we use this there?

File: spec/package_spec.rb (hunk 1 above)
Comment (adammw) [inline spec/package_spec.rb]: @sandlerr ?

File: spec/validations/manifest_spec.rb [hunk 11]
```diff
@@ -1,4 +1,8 @@
 require 'spec_helper'
+require 'tmpdir'
+
+no_code_required_for_iframe_only = 'Javascripts, stylesheets and templates are not '\
```
Comment (adammw) [inline spec/validations/manifest_spec.rb]: Ugh. Can we use I18n.t('no_code_required_for_iframe_only') instead?

File: spec/package_spec.rb (hunk 1 above)
Comment (sandlerr) [inline spec/package_spec.rb]: nope, that require seems unused. `rspec spec/package_spec.rb` worked.

File: spec/fixtures/app_en.js (hunk 6 above)
Comment (sandlerr) [inline spec/fixtures/app_en.js]: done

File: lib/zendesk_apps_support/validations/source.rb (hunk 9 above)
Comment (sandlerr) [inline lib/zendesk_apps_support/validations/source.rb]: fixed

**Comment by adammw** [comment]
:+1:

File: config/locales/en.yml [hunk 12]
```diff
@@ -43,6 +43,7 @@ en:
                 other: 'Missing required oauth fields in manifest: %{missing_keys}'
             missing_source: Could not find app.js
             style_in_template: "<style> tag in %{template}. Use an app.css file instead."
+            no_code_for_ifo_notemplate: Javascripts, stylesheets and templates are not allowed when an iframe URI or noTemplate is specified
```
Comment (mariasica) [inline config/locales/en.yml]: Would it be possible to get a screenshot for translators? What is noTemplate? Written together? Should it be localized or not? Thanks!

Comment (sandlerr) [inline config/locales/en.yml]: `noTemplate` is a special key that people can declare to be true of false. It should be written together and not localised.
